### ISSUE TITLE
Change jsm to WebExtension (jsm deprecated with FF 57)

### ISSUE THREAD

**Issue by AWolf81** [feinstaub/webextension_local_filesystem_links#81]
As mentioned on MDN `jsm` will be deprecated soon and it is not recommended for new Addons - see [here](https://developer.mozilla.org/de/Add-ons/SDK/Tutorials/Getting_Started_(jpm)).

Instead [WebExtensions](https://developer.mozilla.org/de/Add-ons/WebExtensions) should be used and that will also make it compatible with Chrome, Safari (and probably IE). 

WebExtensions are supported with FF releases 51 and above. FF 51.x is released and available.

This is a larger task and it will be done on branch `migration-webextension`.

**Comment by feinstaub** [comment]
I got some info from Mozilla:

**Comment by AWolf81** [comment]
@feinstaub I think we have a deadline for the transfer to webextension. Planned release date for FF 57 is the [14.11.2017](https://calendar.google.com/calendar/render?eid=Y2Ywc2tlM2MxbGE3OThzZGRlbmJ2Zms0ZDggbW96aWxsYS5jb21fMmQzNzM4MzQzMzM1MzQzMjM1MmQzOTM5QHJlc291cmNlLmNhbGVuZGFyLmdvb2dsZS5jb20&ctz=America/Los_Angeles&sf=true&output=xml#eventpage_6).

So please add a deadline to this issue. I'll try to start coding asap.

If you have time you could help me with the native app part that's needed for the web extension. Especially how it could be installed.

I would create the native app as python script becaue that's pretty portable. An installation routine is required because we need to have a registry entry for the [native communication](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Native_messaging). In the demo that I've created it's done with a batch file but that's just OK for development.

**Comment by feinstaub** [comment]
I added a deadline to the next milestone.

Do we really need a custom native app or could we just use an existing one (like windows explorer for windows and xdg-open for Linux)?

**Comment by AWolf81** [comment]
Thanks for adding the milestone.

Windows Explorer and XDG-Open can be used to open the link but for the communication part to the Extension a script is required. I don't like that we need a native app/script but I think there is no other way.

I've modified the example python file from the native communication example app and added the opening logic (tested in Windows).

But at the moment my demo app has an issue with sending the native message. Not sure what's wrong because it was working. Maybe registering the app is not working. I have to do more tests.

Please have a look at my first prototype [here](https://github.com/AWolf81/webextension-local-filelinks-prototype).

Once I'm satisified with the repository of the prototype, I'll transfer the code to a branch here so we're having everything in one place.

**Comment by feinstaub** [comment]
Hi!

1.)
We need communication only for feedback if the file could be started or not, right? So, as fallback we could use this method without a native app, can't we?

2.)
Thanks for providing the prototype.

I followed these instructions https://github.com/AWolf81/webextension-local-filelinks-prototype/blob/master/README.md under openSUSE Linux.

$ npm install
$ npm run dev
http://localhost:8080/

src/host/install_host.sh was not executable. I changed it and run it. Can you fix it in the codebase?

Now, I must admit I don't fully understand what should be done to test the addon in Firefox. Could you add a "for dummies" section in the README.md? :-) Sadly, I am not that familiar with current JavaScript development technologies.

**Comment by AWolf81** [comment]
Hi:

to 1) 
As of my understanding it is always required because web extensions can not access local files. A native app is like the interfacing layer between Firefox and the file explorer. FF tells the native app - hey open that link in Windows explorer or XDG-open and the native app can also return to FF that file access was OK. It would be great if there would be a way with-out it but I think that the chance is small.

to 2)
Why was it not executable? Was just the file attribute +x missing? I'll add that.

Sure, I can add more details to the README.md. But where do you have the problem?
- Can you load the extension in Firefox? There should be an icon in addonbar with a dummy drop down page. Later we can use this for quick access settings.
- Are local file links clickable? This is where I have problems at the moment. It was working to open files but I'm not sure why native communication is currently not working. Or maybe I have trouble with whitepage setting.

## Security concerns
We need to add a hash check that the native app is our installed file at every communication so that a replaced file won't work with the WebExtension. Possible security issue with the extension. A virus/ransomware could use this to trigger an app execution if they find a way to replace our original executable.
Points we need to check here:
- Is it possible to replace the native app with-out changing the WebExtension installation? E.g. user clicks accidentally on an executable file (it will just copy the virus to the native app location) that a hacker created to try if there is our extension to trigger an action. The hacker could directly trigger an action but if it changes the native app it will look like we're doing bad things.
- Executable file check needs same handling as in addon. I've already added this to the Python script but this needs to be verified.
- Checking the native app hash could be difficult. Because how can you trigger a check of the file with-out file access? With file access from extension you could calculate the hash value of the native app. Firefox needs a way to check if the file is unmodified. I couldn't find more details to native messaging security at Google. Maybe we need to chat with some Mozilla Devs to see if there is a way to check the native app or if it is not that critical - but I think it is.

**Comment by feinstaub** [comment]
> to 2)
>Why was it not executable? Was just the file attribute +x missing?

yes

> I'll add that.

thanks

> Sure, I can add more details to the README.md. But where do you have the problem?

I opened the browser - the one I am now browsing with, plus the current Local Filesystem Extension installed - at http://localhost:8080/ and nothing special happens except for seeing the page "Welcome to Your Vue.js App".

> Security concerns

Thanks for making these considerations. We could place your thoughts in an extra issue and/or adding a section to the README.md.

**Comment by MrWillow** [comment]
Hi AWolf81 and feinstaub,

I am curious if you made some progress on this daunting issue.
If I understand everything correctly the upcoming 57 release of firefox (in november) will break this useful add-on completely.

**Comment by AWolf81** [comment]
Hi MrWillow:

I'm sorry there is no progress. I'm pretty busy at the moment and I'm a bit stuck at the prototype. It's probably just an installation issue with the native app but I couldn't find the issue.

Yes, you're right. The release in November will break the add-on and it can't be used anymore.

Thanks for asking and reminding me at the issue. 

I think it should be possible to get it working until November but probably we need to remove some features. So we can focus at the core features - opening links directly, directory open with the icon and whitelisting.

**Comment by feinstaub** [comment]
@AWolf81: thanks for the assessment. As for me, nice-to-have feature removal is ok. The main thing is that the addon is usable at all.

**Comment by feinstaub** [comment]
If you got something I'll test it.

**Comment by MrWillow** [comment]
Good to see a prompt response from you both!
I totally agree: Get something up that works is much more important than nice-to-have features.
If I may suggest:
- Make sure network shares are supported (very common in office like environments)
- whitelisting (with port matching). I saw that things like http://myname:82/ do not work when whitelisted in the latest version. (Where 'myname' isn't something like www.bla.com, but something that resolves locally on the nameserver used)

Good luck! I am looking forward to an updated version.

**Comment by AWolf81** [comment]
@MurzNN Thanks for pointing me to [Open with repo](https://github.com/darktrojan/openwith) in issue #95. I can learn more about how to create webextensions.

I need to check how their installation script is working and how they trigger it. Maybe parts of it can be used here. After some searching in their source code it's clear that they're not having an installer and the user needs to install Python on their own. Also the install script needs to be triggered manually.

An installer is required because otherwise it's too complicated for most of the users to setup the extension. See comments [here](https://addons.mozilla.org/de/firefox/addon/open-with/reviews/).

At least a batch file with a python executable [e.g. with py2exe](http://www.py2exe.org/) that's copied to user's home directory is required to install the native app part.

**Comment by AWolf81** [comment]
Some infos to current look and feel (I think it's OK but if there should be something improved please let me know):
**Addon settings page**
![Addon Settings](https://user-images.githubusercontent.com/3046542/33525727-e2c0c900-d835-11e7-9e26-44cbd5c75cdb.png)

**Test page and popup menu in Addon bar**
![Test page & addon bar icon](https://user-images.githubusercontent.com/3046542/33525614-7fbec1b0-d833-11e7-806b-3287f5d16858.png)
The menu has only a link to the options page (for faster access to the settings) at the moment but I think we can add more feature here later (e.g. add current page to white list or add a link to a help file)

I think it's pretty similar to the previous version just the settings page looks differently but I think it's OK. I'd like to use the css styling from Firefox but wasn't sure how to do it - so I created my own styling.

I'll skip the context menu for now as I think it's optional and we can add it to the next version.

Next I need to create the installer for the native host app for Windows. Once this is done I'll create a package of the extension and post it in the issues. After that I'll move to Unix testing - nothing done yet as I'd like to finish Windows development first.

**Comment by feinstaub** [comment]
Thanks for the status. Looks fine. Some minor remarks:

- The new version would be 0.99.57
- The title on the popup on the menu bar should be consistent to the Extension's name. So the title would be "Local Filesystem Links"

OK to all other items you mentioned.

**Comment by MrWillow** [comment]
Thank you very much. The screenshots look good.
I am looking forward leaving firefox 52.5 ESR behind me. :-)
Keep us posted please.

**Comment by AWolf81** [comment]
I have the extension working under Windows. Installer is ready and working. 
A smaller bug that I need to fix is that the Notifications are not shown at the moment.

Next I need to check Linux and make it work there.

The extension can be tested like following (it's easier to install later - it's just a bit more complicated for the test version):
- Download the [zip file](https://github.com/feinstaub/webextension_local_filesystem_links/blob/migration-webextension/web-ext-artifacts/local_filesystem_links-0.99.57.zip) and extract it to a destination at your machine
- Then open FF & add `about:debugging` as URL + hit enter and load the `manifest.json` in the extracted directory
- Next click on the addon bar icon and click on `show installation guide`
- Now download and install the app with the windows installer (unsigned). You have to click 'show more details in the Windows Defender prompt and click install anyway. (Maybe you have to check install VS redis. at the first dialog if you don't have it - maybe I'll add a check in the installer later).
- The addon is now temporarily installed and can be tested. The app is uninstallable in the Windows program manager.

The install flow should be later like following:
1. Install the extension from Mozilla store
2. Extension shows installation guide
3. Manually install the native app.

@feinstaub Thanks for your remarks. I've added them. 
I have some points that I'm not sure how we should handle them - maybe you're having an idea how to solve:
In the migration branch I've a `dist` folder and an `artifacts` folder. They're both build directories and are created from the `src` directory with npm scripts. What do you think should we add them to the master branch later or should we `.gitignore` these folders? Maybe we could add them in a release branch but I'm not sure. 

What do you think should I create a PR for the migration branch or should I wait until I have it working in Linux?

**Comment by feinstaub** [comment]
> In the migration branch I've a dist folder and an artifacts folder. They're both build directories and are created from the src directory with npm scripts. What do you think should we add them to the master branch later or should we .gitignore these folders?

The build directories should not be part of the repository. They are generated from source; so why should they be checked in? Or what are your thoughts why these folders should be kept?

> Maybe we could add them in a release branch but I'm not sure.

I would not place binary releases into the source repository. One could create a separate repository or use the github downloads feature. Normally, this would not be necessary because the binary release is published on the Mozilla website.

> What do you think should I create a PR for the migration branch or should I wait until I have it working in Linux?

Please go ahead with the Windows-only version. I think it would be good to have a Quantum-ready version as soon as possible. Unless you feel that this step would strongly weaken your future efforts with the Linux version.

**Comment by AWolf81** [comment]
**`/dist` folder topic**
Yes, you're right. The generated files shouldn't be checked-in in source repo. (I'll add dist to `.gitignore`) but I've thought it would be good to have the binaries somewhere. As it could be difficult to generate the data for the Mozilla binary in one build script like it is now.
I think building the extension is the same for all platforms but the native app part will be different for each platform.

Also I'd like to add the native app to the extension bundle, so it's automatically available in the installation guide page. But maybe a link to a download page would be also OK so we could separate the native apps into a separate repository. 

So I think it's required to create a repository for the native apps source in a separate repository and use links to the download location for each platform. Then we can build & bundle the extension on any platform and the native apps can have platform specific build setups.

And one repository for the released binaries for the native apps. So we can use it for downloading the installers. You're right extension binary not needed as it is hosted in Moz. store.

**Finalizing the extension**
I'd like to check Ubuntu if it's not taking too much time. I try to finish this in one or two weeks. I think it has to work at least on Windows and Linux. 
Unit tests can be created for the next release as I think it's OK to manually test the extension for the first release.

@feinstaub Could you please help with the following todos:

- [ ] Check if there is a way to sign the executable files for Windows? I haven't found anything with out paying. I think code signing is required for Windows so that the user knows that the excuteable can be trusted. I'm also not sure if this would be a one time payment or a payment every year.
- [ ] Create the repositories as mentioned for the nativeMessaging host application (source) and one for the released binaries.

**Comment by feinstaub** [comment]
> Check if there is a way to sign the executable files for Windows? I haven't found anything with out paying. I think code signing is required for Windows so that the user knows that the excuteable can be trusted. I'm also not sure if this would be a one time payment or a payment every year.

I skimmed over https://msdn.microsoft.com/en-us/library/ms537361(v=vs.85).aspx "Introduction to Code Signing"

Windows Code signing addresses two issues:
1. "Ensuring authenticity -- Assures users that they know where the code came from."
2. "Ensuring integrity -- Verifies that the code hasn't been tampered with since its publication."

Authenticity: I am not sure if it helps if people are assured that the addon is guaranteed to come from two individuals they don't know anyway. ;-)

Integrity: Do you think it is possible to integrate the native app as binary blob into the addon? Then we could include a download link that points to the integrated file and show instructions of how to install. We would not need to establish and secure a separate download site. Integrity would be established by trusting the Mozilla addon store and the integrity of my Mozilla developer account.

The rest of the document explains a rather complicated infrastructure which I don't fully understand. 

On the other hand, I just found this answer regarding code-signing when dealing with GNU/Linux/FOSS software: https://stackoverflow.com/questions/1732927/signed-executables-under-linux/1734637#1734637. Reading this, I think time is better spent to work towards more free software platforms than to engage in binary code-signing with some minor benefits for a non-free platform. Maybe, in the future, code-signing will be mandatory for Windows anyway. I will wait until then.

Are there any good reasons we should do the code-signing as soon as possible?

> Create the repositories as mentioned for the nativeMessaging host application (source) and one for the released binaries.

Since the addon only works together with the native messaging host application I think we should keep this code together with the addon's code. Maybe rename the "host" folder to "native-host-application" and move one level up. Then have subfolders "windows-src", "windows-bin", "linux-src", "linux-bin" which contains the source and binaries. In this case it would make sense to me to include the binaries. The *-bin folders would be the source for bundling the native host app with the extension package. We could bundle all platform specific binaries into one extension package. What do you think?

**Comment by AWolf81** [comment]
**Code signing**
Thanks for your research. Yes, you're probably right with authenticity. :-)
I think that we can not bundle the installer into the extension so that it is not showing the warning. At the moment the installer is part of the extension and after clicking on download link you have the file locally and after starting you're getting the warning. (Maybe if it's possible to directly open the executable from the extension it won't show the message but I'm not sure if this is possible.)

I think we don't have to sign the files but it would be better. The installer can be started even it is unsigned but the warning message is a bit confusing.

The topic is really not that easy to fix. So I think we could fix this later. But we should keep this in mind and solve as soon as possible because nobody likes to click away system warning messages and install it anyway.

**Host folder**
Good point. That's also OK and probably easier to manage. This way we can also add the download link to the installer files in the binary directory from Github. The source of the native app seems to be cross-platform as it is a Python script. But the installers will be different. I think it's OK like this (starting in top-level directory):

    /native-host
       package.json (not sure if it's required - I'll check if this is working for build scripts)
       /src (contains the Python script)
       /build
         /windows (build related files)
         /linux
       /bin/windows (installation files)
       /bin/linux

I'll create the directory structure like this.

**Comment by AWolf81** [comment]
Fixed with 0.99.57

**Comment by MurzNN** [comment]
Please provide in Extension description text on Mozilla website - info that it need to install external application, and link to page with instruction how to install native application.
